rfcs/20210624-persistent-cache/README.md

```diff
@@ -0,0 +1,259 @@
+is created from scratch
+Option 2: Binary representations that are in the primitive cache at the time of
+primitive cache destruction are stored in the persistent cache
+
+The proposal is to go with the option 1 to avoid unnecessary dependency between
+persistent cache and primitive cache. This is expected that primitive cache miss
+doesn't occur often in workloads otherwise the performance will be affected either
+way.
+
+#### Interprocess and Thread Synchronization
+
+The access to the persistent cache should be performed in a process and thread
+safe manner. For performance reasons both process and thread synchronization
+should be performed using readers-writer lock.
+
+### Option 2: Provide API for Implementing Persistent Cache
+
+oneDNN provides an API so that users can implement persistent cache in their
+applications.
+
+#### API
+
+The API will be responsible for providing a user with an ID for the binary
+representation of a kernel and the binary representation itself.
+
+* ID is an array of bytes represented as `std::vector<unsigned char>` that is a
+unique identifier for a binary representation. ID should take into account that
+binaries compiled for different devices are different binaries. The binaries
+compiled for a sub-device considered compatible with the main device.
+* Binary representation is an array of bytes represented as `std::vector<unsigned char>`.
+Since oneDNN primitive can consist of multiple kernels the API return
+`std::vector<std::vector<unsigned char>>`.
+
+This proposal covers only providing binary representations of kernels. However,
+it's possible that oneDNN will support serialization of the primitives in the future
+so that users can cache the whole model on disk. Given that, it makes sense to
+make API more common and scalable.
+
+The suggestion is to introduce a binary kind to distinguish binary representation of
+a kernel from primitive.
+
+```cpp
+struct primitive {
+    enum class binary_kind {
+        kernel,
+        /*primitive - unimplemented */
+    };
+    /* ... */
+};
+```
+
+The API for obtaining ID will be provided by the `primitive_desc_base` class
+therefore the API will be available in each `primitive_desc` class.
+
+```cpp
+struct primitive_desc_base {
+    /* ... */
+    std::vector<unsigned char> get_id() const;
+};
+```
+
+The `primitive` class will get a new constructor to take a binary and new API
+to provide user with the binary.
+
+```cpp
+struct primitive {
+    /* ... */
+    primitive(const primitive_desc &pd, const std::vector<std::vector<unsigned char>> &binaries, binary_kind kind);
+    // Primitives that are not supported return an empty vector.
+    std::vector<std::vector<unsigned char>> get_binaries(binary_kind kind) const;
+};
+```
+
+##### API usage example
+
+```cpp
+using namespace dnnl;
+
+{
+    convolution_forward::primitive_desc conv_pd(desc, attr, engine);
+    primitive conv(conv_pd);
+
+    auto key = conv_pd.get_id();
```

Comment:
What do we do with zero pad? Do we introduce memory.get_binaries(memory::binary_kind::kernel) as well?

Reply:
Can you please elaborate? I'm not sure I understand how memory is relevant to the binary.

Reply:
As far as I remember we have an OpenCL kernel for zero pad, which we sometimes use to pad memory on GPU.

Reply:
I see, we have such a kernel for memory but the kernel is universal hence doesn't depend on memory descriptor. Therefor it will be created only 1 time for the entire execution. I Because of that, I think that adding such API for memory is an overkill. I also think we can re-write the kernel to nGEN to speed up its creation, this will be possible after OpenCL starts supporting zebin binary format.

Reply:
makes sense, let's keep it out of this RFC until it is a bottleneck. At lease we have a way to address the issue in the future.